### PR TITLE
Implement inline generics

### LINKED ISSUE #56: Support `inline` and `flatten` for generic types
Currently, inlining fails to take the used generic type argument into account:
```rust
struct X<T> {
  t: T
}
struct Y {
  #[ts(inline)]
  x1: X<i32>,
  #[ts(flatten)]
  x2: X<i32>
}
```

### PR DESCRIPTION
**Pull request by escritorio-gustavo**
Implementation of inlined generics. Not 100% sure it works, but it did pass the `#[ignore]`d test

Fixes #56

### CODE REVIEW

**Comment by escritorio-gustavo** [comment]
This may have a flaw for the case where generics have both a trait bound and a default type, in which case it will keep the previous behavior. Not sure how to avoid this though

File: macros/src/types/generics.rs [hunk 1]
```diff
@@ -55,9 +55,20 @@ pub fn format_type(ty: &Type, dependencies: &mut Dependencies, generics: &Generi
...
+        if !generic.bounds.is_empty() || generic.default.is_some() {
+            return quote!(#generic_ident_str.to_owned());
+        }
+
+        return quote!(
+            match <#generic_ident>::name().as_str() {
+                "null" => #generic_ident_str.to_owned(),
+                x => x.to_owned()
+            }
+        );
```
[the remark this answers is not in the document]
Comment (NyxCode) [reply macros/src/types/generics.rs:71]: Ah, okay, I understand! Let me quickly check an edge-case I suspect breaks here.

[the remark this answers is not in the document]
Comment (NyxCode) [reply macros/src/types/generics.rs:71]: A small comment here might be helpfull to clarify why "null" is here, that took my quite a bit to figure out.

**Review by NyxCode — APPROVED** [review]
Besides maybe adding a small comment why there's this `"null"` literal, this looks great!
Otherwise, this looks great, thanks!

File: macros/src/types/generics.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (NyxCode) [reply macros/src/types/generics.rs:71]: > Ah, okay, I understand! Let me quickly check an edge-case I suspect breaks here.

Nevermind, works nicely! :D

**Comment by escritorio-gustavo** [comment]
> to make exporting generic types with T::<()>::decl() work, right?
Yes, when exporting the struct itself, `()` is used as the default type, which turns into `"null"`. When you give it a type argument in a struct field, it uses the type's `.name()`

**Comment by escritorio-gustavo** [comment]
> This may have a flaw for the case where generics have both a trait bound and a default type
The reason for this is a test that has something along the lines of `Struct<T: ToString>`. In this case, the default type is not `()`, it is `&'static str`. I don't know how the default type is decided in this case, so I wasn't able to implement that, and even if I did, if someone were to use the default type (or any type where `.name()` has the same result) this implementation would just give them T

**Comment by NyxCode** [comment]
> > This may have a flaw for the case where generics have both a trait bound and a default type
> 
> The reason for this is a test that has something along the lines of `Struct<T: ToString>`. In this case, the default type is not `()`, it is `&'static str`. I don't know how the default type is decided in this case, so I wasn't able to implement that, and even if I did, if someone were to use the default type (or any type where `.name()` has the same result) this implementation would just give them `T` (or whatever the name of the type param is)

I think that's fine for now, this PR is still an improvement on the status quo. Maybe we should create an issue for this, together with an ignored test case, so that we dont forget about this?

**Comment by escritorio-gustavo** [comment]
That might be the best way to go. Btw, I tried an aproach using `std::any::Any` and `std::any::TypeId`. That seems like the best way to check if two Rust types are equal, but I ran into lifetime problems (`T` is required to be `'static`) so I went for this approach. Maybe that will be useful in the future?

**Comment by escritorio-gustavo** [comment]
> This may have a flaw for the case where generics have both a trait bound and a default type, in which case it will keep the previous behavior. Not sure how to avoid this though

This actually happens in case **either** a default type **or** a trait bound exist

**Comment by escritorio-gustavo** [comment]
The best way to solve this permanently would be a way to identify whether we are expanding a `flatten`/`inline` or an `export`.
`flatten` and `inline` should always emit `T::name()`
`export` should always emit `T`

**Pull request merged**